core/src/services/compfs/writer.rs

```diff
@@ -44,27 +48,35 @@ impl oio::Write for CompfsWriter {
     /// the write_all doesn't work correctly if `bs` is non-contiguous.
     ///
     /// The IoBuf::buf_len() only returns the length of the current buffer.
-    async fn write(&mut self, bs: Buffer) -> Result<()> {
-        let mut file = self.file.clone();
+    async fn write(&mut self, mut bs: Buffer) -> Result<()> {
+        let mut file = self.file.clone().unwrap();
 
-        self.core
+        let pos = self
+            .core
             .exec(move || async move {
-                buf_try!(@try file.write_all(bs).await);
-                Ok(())
+                while bs.has_remaining() {
```

Comment:
`Buffer` implements `Iterator<Item=Bytes>`, we can use `for b in bs` here to avoid extra copy of data.